evennia/web/webclient/static/webclient/js/webclient_gui.js

```diff
@@ -76,36 +96,91 @@ function doSendText() {
 // catch all keyboard input, handle special chars
 function onKeydown (event) {
     var code = event.which;
-    inputfield = $("#inputfield");
+    var history_entry = null;
+    var inputfield = $("#inputfield");
     inputfield.focus();
 
     if (code === 13) { // Enter key sends text
         doSendText();
         event.preventDefault();
-    }
-    else if (code === 38) { // Arrow up
-        inputfield.val(input_history.back());
+    } 
+    else if (inputfield[0].selectionStart == inputfield.val().length) { 
+        // Only process up/down arrow if cursor is at the end of the line.
+        if (code === 38) { // Arrow up
+            history_entry = input_history.back();
+        }
+        else if (code === 40) { // Arrow down
+            history_entry = input_history.fwd();
+        }
+    } 
+    
+    if (history_entry !== null) {
+        // Doing a history navigation; replace the text in the input.
+        inputfield.val(history_entry);
         event.preventDefault();
     }
-    else if (code === 40) { // Arrow down
-        inputfield.val(input_history.fwd());
-        event.preventDefault();
+    else {
+        // Save the current contents of the input to the history scratch area.
+        setTimeout(function () {
+            // Need to wait until after the key-up to capture the value.
+            input_history.scratch(inputfield.val());
+            input_history.end();
+        }, 0);
     }
 };
 
+var resizeInputField = function () {
+    var min_height = 50;
+    var max_height = 300;
+    var prev_text_len = 0;
+    
+    // Check to see if we should change the height of the input area
+    return function () {
+        var inputfield = $("#inputfield");
+        var scrollh = inputfield.prop("scrollHeight");
+        var clienth = inputfield.prop("clientHeight");
+        var newh = 0;
+        var curr_text_len = inputfield.val().length;
+        
+        if (scrollh > clienth && scrollh <= max_height) {
+            // Need to make it bigger
+            newh = scrollh;
+        } 
+        else if (curr_text_len < prev_text_len) {
```

Comment:
This looks like an interesting solution! Out of curiosity, is using an offscreen element really the only way to measure the size of a text? Shouldn't this somehow be dependent on inputsizer having the same css properties as the field we are trying to measure as well (it doesn't look like it does)? Not saying it's wrong, just confused about how it works. :)